orion-server/src/main/java/com/pinterest/orion/core/actions/aws/AmiTagManager.java

```diff
@@ -0,0 +1,81 @@
+package com.pinterest.orion.core.actions.aws;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import com.pinterest.orion.server.api.AMI;
+import com.pinterest.orion.server.api.BaseClustersApi;
+
+import software.amazon.awssdk.services.ec2.Ec2Client;
+import software.amazon.awssdk.services.ec2.model.DescribeImagesRequest;
+import software.amazon.awssdk.services.ec2.model.DescribeImagesResponse;
+import software.amazon.awssdk.services.ec2.model.CreateTagsRequest;
+import software.amazon.awssdk.services.ec2.model.CreateTagsResponse;
+import software.amazon.awssdk.services.ec2.model.Filter;
+import software.amazon.awssdk.services.ec2.model.Tag;
+
+public class AmiTagManager {
+  private static final Logger logger = Logger.getLogger(AmiTagManager.class.getCanonicalName());
+  private Ec2Client ec2Client;
+
+  public AmiTagManager() {
+    ec2Client = Ec2Client.create();
+  }
+
+  public List<AMI> getAmiList(Map<String, String> filter) {
+    List<AMI> amiList = new ArrayList<>();
+    DescribeImagesRequest.Builder builder = DescribeImagesRequest.builder();
+    builder = builder.filters(
+      Filter.builder().name("tag:application").values("kafka").build()
+    );
+    if (filter.containsKey("os"))
+      builder = builder.filters(
+        Filter.builder().name("tag:release").values(filter.get("os")).build()
+      );
+    if (filter.containsKey("arch"))
+      builder = builder.filters(
+        Filter.builder().name("tag:cpu_architecture").values(filter.get("arch")).build()
+      );
+    builder = builder.filters(
+      Filter.builder().name("tag:application_environment").values("*").build()
+    );
+    DescribeImagesResponse resp = ec2Client.describeImages(builder.build());
+    if (resp.hasImages() && !resp.images().isEmpty()) {
```

Comment:
Just curious, is this possible to be false? If so, what's the case?

Reply:
It's false if no images are retrieved. Even when `resp.hasImages()` is true, there's still a chance that `resp.images()` is an empty list. And we can't check `resp.images()` only, as it's null when `resp.hasImages()` is false. Hence the double check. Looks like a poor implementation of the AWS Java SDK.